packages/sanity/src/core/studio/copyPaste/useCopyPasteAction.ts

```diff
@@ -0,0 +1,333 @@
+            }),
+          })
+          return
+        }
+
+        const targetSchemaTypeTitle = targetSchemaType.title || targetSchemaType.name
+        const transferValueOptions = {
+          sourceRootSchemaType: sourceSchemaType,
+          sourcePath: [],
+          sourceValue: item.value,
+          targetRootSchemaType: targetSchemaType,
+          targetPath: [],
+          // This will mainly be used for validating references with filter callback that
+          // needs the document and absolute path to the field
+          targetRootPath: targetPath,
+          targetRootValue: value,
+          options: {
+            validateAssets: true,
+            validateReferences: true,
+            client,
+          } as TransferValueOptions,
+        }
+        copiedJsonTypes.push(sourceSchemaType.jsonType)
+
+        try {
+          const {targetValue, errors} = await transferValue(transferValueOptions)
+          const nonWarningErrors = errors.filter((error) => error.level !== 'warning')
+          const _isEmptyValue = isEmptyValue(targetValue)
+
+          if (nonWarningErrors.length > 0) {
+            const description = t(nonWarningErrors[0].i18n.key, nonWarningErrors[0].i18n.args)
+
+            toast.push({
+              status: 'error',
+              title: t('copy-paste.on-paste.validation.clipboard-invalid.title'),
+              description,
+            })
+            return
+          }
+
+          if (errors.length > 0 && !_isEmptyValue) {
+            const description = errors.map((error) => t(error.i18n.key, error.i18n.args)).join(', ')
+
+            toast.push({
+              status: 'warning',
+              title: t('copy-paste.on-paste.validation.partial-warning.title'),
+              description,
+            })
+          }
+
+          if (_isEmptyValue) {
+            toast.push({
+              status: 'warning',
+              title: t('copy-paste.on-paste.validation.clipboard-empty.title'),
+            })
+            return
+          }
+
+          updateItems.push({patches: [set(targetValue, targetPath)], targetSchemaTypeTitle})
+        } catch (error) {
+          toast.push({
+            status: 'error',
+            title: error.message,
+          })
+          return
+        }
+      }
+
+      telemetry.log(FieldPasted, {
+        context: options?.context?.source || 'unknown',
+        schemaTypes: copiedJsonTypes,
+      })
+
+      if (updateItems.length) {
+        const allPatches = flatten(updateItems.map(({patches}) => patches))
+        const allTargetNames = updateItems
+          .map(({targetSchemaTypeTitle}) => targetSchemaTypeTitle)
+          .join('", "')
+
+        if (!onChange) {
+          console.warn(`Failed to resolve document onChange method when pasting.`)
+
+          toast.push({
+            status: 'error',
+            title: t('copy-paste.on-copy.validation.document-metadata-unknown-error.title'),
+          })
+          return
+        }
```

Comment:
This happened earlier because of a race condition where the callback sometime wouldn't update because the `documentMeta` was coming from a dependency that itself wouldn't trigger a rerender. I believe it is stable now.